### ISSUE TITLE
Catch discrepancies between versions (in CMakeLists.txt and debian/changelog)

### ISSUE THREAD

**Issue by AlexisWilke** [Zipios/Zipios#27]
The version is found in two places:

* `debian/changelog` -- the Debian changelog with comments of the changes to that version
* `CMakeLists.txt` -- the cmake file which updates the version in the .in files (i.e. `zipios/zipios-config.hpp.in`)

We need a way to detect whether there is a discrepancy between the versions.

The simplest, I think, at this time is to have a test which compares the versions. I don't always run the tests, but I would imagine that can help already. That test could be run in the standard build process, though, that way it would quickly be detected since I run builds often.

**Comment by sevu** [comment]
And thirdly the git version number. So, let's give it a try:
- CmakeLists.txt: you wrote already the script `dev/version`, if one sources it, one has the environment variable
- debian/changelog: `head -1  debian/changelog | cut -f 2 -d "(" | cut -f 1 -d "~"` extractes the last version number
- git: `git describe --tags --abbrev=0 | sed s/^v//` gives the version of the last tag in a comparable format. `--tags` to handle both lighteight and annotated tags (or better use annotated tags), `--abrev=0` to give the number of the last tag, ignoring the unreleased commits.

Now, how to make a test out of it / where to put it? Extending `dev/version`?

**Comment by AlexisWilke** [comment]
Yeah that would probably work. All we'd have to do is add a target which always runs that script when we do a `make`. Then it should detect the discrepancy and report it with an error breaking the `make` and thus forcing the developer to fix the version in one place or another.

In regard to the `git` info, are you thinking that each version should be tagged?

**Comment by dooglio** [comment]
Another way is to add a `changelog.in`, which is replaced via CMake itself and allows the `CMakeLists.txt` file itself to drive the version (using [`configure_file`](https://cmake.org/cmake/help/v3.0/command/configure_file.html)).

**Comment by AlexisWilke** [comment]
I don't think that a `changelog.in` would work with `debuild -S -sa` which we use to send projects to launchpad. I thought of that a while back...

Now, in our other Snap! project we make use of the "get version" like so:

    SnapGetVersion( SNAPWEBSITES ${CMAKE_CURRENT_SOURCE_DIR} )

and that cmake function retrieves the version from the `debian/changelog` file. In other words, a single location with the project version.

Only `zipios` doesn't use our `cmakesnapwebsites` module (to remain independent) so that function is not available.

I suppose we could re-implement that function in `zipios` using sevu's script (`head -1 ...`). That is, we would convert our existing `dev/version` script using sevu's script and then have cmake call the script and stick that version in variables. Using sevu's script means it would work on _all_ unices at least (whereas, the existing function only works on computers with the Debian script we run to extract the version.)

There is also the problem of MS-Windows and it's total lack of sane command line tools... (I know the library is not yet really working under MS-Windows, but we probably shouldn't completely prevent that move.)

**Comment by dooglio** [comment]
>There is also the problem of MS-Windows and it's total lack of sane command line tools...

That is why I feel that `CMakeLists.txt` should drive the version, *not* `debian/changelog`. The package should be generated on build, and the `changelog` file a purely auto-generated file.

**Comment by AlexisWilke** [comment]
Could we not write a function in a CMake module which reads the first line from the changelog file and extract the version? I would think CMake can handle regex/string functions so we should be able to handle that with pure CMake code, right? Then it would work under MS-Windows too.

    zipios (2.2.1.0~xenial) xenial; urgency=high

That's something easy to parse in CMake right? Create a list of words, pick the second one, remove the first/last characters (parenthesis) and then deal with the `~<name>` part...

**Comment by dooglio** [comment]
It makes sense for a project like `SnapC++`, which is intended for Linux only.  Since `ZipIOS` is designed to be cross-platform,  I think it should be the `CMakeLists.txt` file that drives it, and `changelog` should reflect that instead.
